File: test/unroll-vectorization/gauss/gauss_elimination.f

      program gauss_elimination
          implicit none

          integer, parameter :: n  = 4096 

          double precision, dimension(n,n) :: A
          double precision, dimension(n)   :: b, x

          integer :: i, j
          do  i = 1, n 
          b(i) = i*0.1d1 + 1.0d2
          do j = 1, n
          A(j,i) = i*0.7d1 -j*0.33d1 + 1.0d1
          enddo
          enddo

          x = gauss( A, b, n )

      contains
          ! Elimination de Gauß pour triangulariser un système linéaire
          function gauss ( A, b, n )
              implicit none

              double precision, dimension(n, n), intent(inout) :: A
              double precision, dimension(n),    intent(inout) :: b
              integer, intent(in) :: n

              double precision, dimension(n) :: gauss 
              double precision :: m 
              integer :: k, i, j

              do k = 1, n-1
              do i = k+1, n
              m = A(i,k) / A(k,k)
              b(i) = b(i) - b(k)

              do j = k+1, n
              A(j,i) = A(j,i) - m * A(k,i)
              enddo
              enddo
              enddo

              gauss = remontee(A,b,n)

          end function gauss

          ! Méthode de remontée pour résoudre un système triangularisé
          function remontee ( U, b, n )

              implicit none

              double precision, dimension(n, n), intent(in) :: U
              double precision, dimension(n),    intent(in) :: b
              integer, intent(in) :: n

              double precision, dimension(n) :: remontee
              integer :: i, j

              remontee(n) = b(n) / U(n, n)

              do i = n-1, 1, -1
              do j = i, n
              remontee(i) = (b(i) - U(i, j) * remontee(j)) / U(i,i)
              enddo
              enddo

              return

          end function remontee

      end program gauss_elimination

File: sequential/unroll-vectorization/gauss/gauss_elimination.c

#include <stdlib.h>

// Methode de remontee pour résoudre un système triangulaire
double *
Back_Substitution (const double *restrict U, const double *restrict b,
                   const size_t n)
{

    void *x = NULL;
    posix_memalign (&x, 32, n * sizeof (double));
    double *p_x = __builtin_assume_aligned (x, 32);
    double *p_U = __builtin_assume_aligned (U, 32);
    double *p_b = __builtin_assume_aligned (b, 32);

    p_x[n - 1] = p_b[n - 1] / p_U[(n - 1) * n + (n - 1)];

    for (long long i = n - 2; i > 0; i--)
        {
            for (size_t j = i; j < n; j++)
                {
                    p_x[i]
                        = (p_b[i] - p_U[i * n + j] * p_x[j]) / p_U[i * n + i];
                }
        }

    return p_x;
}

// Elimination de Gauß pour triangulariser un système linéaire
double *
Gauss_Elimination (double *restrict A, double *restrict b, const size_t n)
{

    double *p_A = __builtin_assume_aligned (A, 32);
    double *p_b = __builtin_assume_aligned (b, 32);

    for (size_t k = 0; k < n - 1; k++)
        {
            for (size_t i = k + 1; i < n; i++)
                {

                    const double m = p_A[i * n + k] / p_A[k * n + k];
                    p_b[i] -= m * p_b[k];

                    for (size_t j = k + 1; j < n; j++)
                        {

                            p_A[i * n + j] -= m * p_A[k * n + j];
                        }
                }
        }

    return Back_Substitution (p_A, p_b, n);
}

int
main (void)
{

    // La taille est susceptible d'être changé pour avoir un temps d'exécution
    // assez long pour LProf
    const size_t n = 4096;
    void *A = NULL, *b = NULL;

    posix_memalign (&A, 32, n * n * sizeof (double));
    posix_memalign (&b, 32, n * sizeof (double));

    double *p_A = __builtin_assume_aligned (A, 32);
    double *p_b = __builtin_assume_aligned (b, 32);

    // Remplissage de la matrice et du vecteur
    for (size_t i = 0; i < n; i++)
        {
            p_b[i] = i * 0.1 + 1;
            for (size_t j = 0; j < n; j++)
                {
                    p_A[i * n + j] = i * 0.7 - j * 0.33 + 1;
                    if (p_A[i * n + j] == 0.0)
                        p_A[i * n + j] = 1.0;
                }
        }

    // Appel de la fonction
    double *x = Gauss_Elimination (p_A, p_b, n);

    free (p_A);
    free (p_b);
    free (x);

    return 0;
}
